### DISCUSSION TITLE
How to set an IP address to vm

### DISCUSSION THREAD

**Discussion by thakkarjigneshX** [vmware/govmomi#2927] [Community Help]
Hello,
I created the datacenter, cluster, datastore and vm using below commands.
However, when I check the vm ip, it shows empty. How can I populate the same? What am I missing?

```
govc datacenter.create TestDC
govc cluster.create -dc=TestDC TestCluster
govc cluster.add -dc=TestDC -hostname x.x.x.x -username user -password pass -noverify
govc datastore.create -dc=TestDC -type local -name localDatastore111 -path /var/local x.x.x.x
govc vm.create -c 2 -m 10000 -disk 200GB -net.adapter e1000 -disk.controller pvscsi -dc=TestDC TestVM 
govc vm.info -dc=TestDC TestVM
Name:           TestVM
  Path:         /TestDC/vm/TestVM
  UUID:         88aeeca4-87a6-5b24-8e05-fac539959a2e
  Guest name:   otherGuest
  Memory:       10000MB
  CPU:          2 vCPU(s)
  Power state:  poweredOn
  Boot time:    2022-08-15 04:59:50.317078979 -0700 PDT
  IP address:
  Host:         10.46.213.108
nutanix@NTNX-diagnostics:~$

```

Also, govc metric.ls -dc=TestDC TestVM shows the list of available metrics.
Are these metrics getting auto-populated or we can populate a few/some of them for a given vm?

**Comment by thakkarjigneshX** [comment]
Any comment or help on this?

**Comment by tenthirtyam** [comment]
See [vm.create](https://github.com/vmware/govmomi/blob/master/govc/USAGE.md#vmcreate) and [vm.customize](https://github.com/vmware/govmomi/blob/master/govc/USAGE.md#vmcustomize).

**Ryan Johnson**
Senior Staff Solutions Architect | Product Engineering @ VMware, Inc.

**Comment by tenthirtyam** [comment]
Duplicate of https://github.com/orgs/vmware/discussions/2930.

**Ryan Johnson**
Senior Staff Solutions Architect | Product Engineering @ VMware, Inc.

**Comment by thakkarjigneshX** [comment]
@tenthirtyam , thanks for reply.
Below command will get the mac assigned. How about IP?

-net=                  Network [GOVC_NETWORK]
-net.adapter=e1000     Network adapter type
-net.address=          Network hardware address

**Comment by tenthirtyam** [comment]
Use `vm.customize`.

**Ryan Johnson**
Senior Staff Solutions Architect | Product Engineering @ VMware, Inc.

**Comment by thakkarjigneshX** [comment]
I tried that as well. Getting NicSettingMismatch error. Any help on how to correct this? 

```
[root@diagnostics ~]# govc vm.power -off DC0_H0_VM0
Powering off VirtualMachine:vm-54... OK
[root@diagnostics ~]# govc vm.customize -vm DC0_H0_VM0 -mac 00:50:56:be:dd:f8 -ip 10.0.0.178 -netmask 255.255.255.0 -mac 00:50:56:be:60:cf -ip 10.0.0.162 -netmask 255.255.255.0
govc: *types.NicSettingMismatch
[root@NTNX-diagnostics ~]#

**Comment by tenthirtyam** [comment]
Seems related to https://vdc-download.vmware.com/vmwb-repository/dcr-public/da47f910-60ac-438b-8b9b-6122f4d14524/16b7274a-bf8b-4b4c-a05e-746f2aa93c8c/doc/vim.fault.NicSettingMismatch.html.

Does the template use e1000 or vmxnet3?

**Comment by thakkarjigneshX** [comment]
I just tried following..But don't see the IP assigned.

```
[root@diagnostics ~]# govc ls /DC0/network
/DC0/network/VM Network
/DC0/network/DVS0
/DC0/network/DVS0-DVUplinks-9
/DC0/network/DC0_DVPG0
[root@diagnostics ~]# govc vm.network.add -vm DC0_H0_VM0 -net=DVS0-DVUplinks-9
[root@diagnostics ~]# govc vm.customize -vm DC0_H0_VM0 -mac 00:50:56:be:dd:f8 -ip 10.0.0.178 -netmask 255.255.255.0 -mac 00:50:56:be:60:cf -ip 10.0.0.162 -netmask 255.255.255.0
[root@diagnostics ~]#

[root@diagnostics ~]# govc vm.info DC0_H0_VM0
Name:           DC0_H0_VM0
  Path:         /DC0/vm/DC0_H0_VM0
  UUID:         265104de-1472-547c-b873-6dc7883fb6cb
  Guest name:   otherGuest
  Memory:       32MB
  CPU:          1 vCPU(s)
  Power state:  poweredOn
  Boot time:    2022-08-23 09:37:39.222275523 -0700 PDT
  IP address:
  Host:         DC0_H0
[root@diagnostics ~]#

**Comment by thakkarjigneshX** [comment]
Looks like a bug on vm.info output. Assigned IP gets reflected when I try to use vm.ip.

```
[root@diagnostics ~]# govc vm.ip -n 00:50:56:be:dd:f8 DC0_H0_VM0
10.0.0.178
[root@diagnostics ~]#

**Comment by thakkarjigneshX** [comment]
Tried creating VM with net.adapter `vmxnet3` and then customize to add IP and mac. Still no luck!

```
[root@diagnostics ~]# govc vm.create -c 2 -m 10000 -disk 200GB -g freebsd64Guest -net.adapter vmxnet3 -disk.controller pvscsi -net DVS0-DVUplinks-9 -pool /DC0/host/DC0_H0/Resources Test1
[root@diagnostics ~]#
[root@diagnostics ~]# govc vm.power -off Test1
Powering off VirtualMachine:vm-81... OK
[root@diagnostics ~]# govc vm.customize -vm Test1 -mac 00:50:56:be:dd:f8 -ip 10.0.0.178 -netmask 255.255.255.0
[root@diagnostics ~]# govc vm.power -on Test1
Powering on VirtualMachine:vm-81... OK
[root@diagnostics ~]#
[root@diagnostics ~]# govc vm.info Test1
Name:           Test1
  Path:         /DC0/vm/Test1
  UUID:         7f18204c-961f-5685-961b-db1a011da3c9
  Guest name:   freebsd64Guest
  Memory:       10000MB
  CPU:          2 vCPU(s)
  Power state:  poweredOn
  Boot time:    2022-08-23 09:52:09.390729183 -0700 PDT
  IP address:
  Host:         DC0_H0
[root@diagnostics ~]#

**Comment by tenthirtyam** [comment]
Depends on the template. For example, is it configured for cloud-init by default (eg Ubuntu 20.04 and later) or does it support guest customization (and have the necessary pre-reqs - like RHEL/derivative + Perl)

Do you see it failing in the VM's tasks?

**Comment by thakkarjigneshX** [comment]
Task is not failing.

```
CreateVm                                 group-3                        vcsim                           10:23:34  10:23:34  10:23:34 success [1.936257ms]
PowerOn                                  vm-104                         vcsim                           10:23:34  10:23:34  10:23:34 success [481.414µs]
PowerOff                                 vm-104                         vcsim                           10:23:41  10:23:41  10:23:41 success [383.079µs]
CustomizeVm                              vm-104                         vcsim                           10:23:52  10:23:52  10:23:52 success [138.532µs]
PowerOn                                  vm-104                         vcsim                           10:23:58  10:23:58  10:23:58 success [399.02µs]

**Comment by tenthirtyam** [comment]
Can you confirm that you see it running the Guest Customization under **Monitor** > **Tasks** and that it successfully performed the _Reconfiguring..._ and is not passing the IP configuration?

Can you confirm the guest OS and that running a guest customization the traditional method also does not fail on the same template?

More information is needed for community assistance.

(Note: I won't be back on this evening.)
